### PR TITLE
feat(stats): add a stat recording cubic's w_max

### PR DESCRIPTION
**Pull request by omansfeld**
I would like to use this as an approximation for the ideal congestion window size to compare it with `slow_start_exit_cwnd` and rate exit-point accuracy. 

We already compare with `cwnd` in the new `slow_start_exit_accuracy` glean probe being added in https://bugzilla.mozilla.org/show_bug.cgi?id=2021867 and I would like to have this as an alternative, to confirm if the two yield similar results.

~~Depends on #3278. I will rebase once that lands on main. Only commit belonging to this PR is dd20486ea6f871efd36d181dd089774a9a3058e1.~~

### CODE REVIEW

File: neqo-transport/src/stats.rs [hunk 1]
```diff
@@ -148,6 +148,11 @@ pub struct CongestionControlStats {
     /// None if we haven't exited slow start or if we re-entered after spurious congestion.
     /// When exiting via congestion event, this is the cwnd AFTER the reduction.
     pub slow_start_exit_cwnd: Option<usize>,
+    /// Cubic's `w_max`: the congestion window (in bytes) just before the most recent
+    /// congestion reduction (with fast convergence applied). Reports 0 (the initial value) if no
+    /// congestion event has occurred or Cubic is not in use. Recorded as a stat to approximate a
+    /// connection's ideal congestion window in metrics.
+    pub w_max: usize,
```
Comment (larseggert) [inline neqo-transport/src/stats.rs]: Is it is an `f64` internally, why not export as such?

Comment (omansfeld) [reply neqo-transport/src/stats.rs]: Glean cannot take floats, that was the initial reason. On second thought It makes more sense to do the conversion on the Firefox side though, and keep Neqo coherent with itself. Will change to f64.

File: neqo-transport/src/cc/cubic.rs [hunk 2]
```diff
@@ -472,5 +477,7 @@ impl WindowAdjustment for Cubic {
             self.current
         );
         self.current = stored;
+        // The stat is `None` if it is unset, so set it to that instead of `Some(0.0)`.
+        cc_stats.w_max = (self.current.w_max != 0.0).then_some(self.current.w_max);
```
Comment (omansfeld) [inline neqo-transport/src/cc/cubic.rs]: @larseggert I made this stat an `Option` too, as per your suggestion. As you can see the actual Cubic `w_max` is also using `0.0` as it's *has not been set*-value, prompting the above line of code to align stat and reality. Should `Cubic::w_max` also be changed to an `Option` or is that overkill (for this PR)?

Comment (larseggert) [reply neqo-transport/src/cc/cubic.rs]: I think it would be cleaner if it were. Also allows distinguishing "never set" from "happens to be zero" (which isn't really possible I guess but anyway.)

**Comment by omansfeld** [comment]
@larseggert can you take another now that I also made `Cubic::w_max` an `Option` in https://github.com/mozilla/neqo/pull/3431/commits/d84b8ea632ccef85b57256da874867f1af0b34a3?

File: neqo-transport/src/cc/tests/cubic.rs [hunk 3]
```diff
@@ -304,7 +304,7 @@ fn congestion_event_slow_start() {
     // w_max is equal to cwnd before decrease.
     let cwnd_initial_f64 = convert_to_f64(cubic.cwnd_initial());
     assert_within(
-        cubic.congestion_control().w_max(),
+        cubic.congestion_control().w_max().unwrap(),
         cwnd_initial_f64 + convert_to_f64(cubic.max_datagram_size()),
```
Comment (larseggert) [inline neqo-transport/src/cc/tests/cubic.rs:308]: See https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/ for why using Epsilon in this way is flaky.

File: neqo-transport/src/cc/classic_cc.rs [hunk 4]
```diff
@@ -1623,6 +1625,8 @@ mod tests {
         assert_eq!(cc_stats.congestion_events[CongestionEvent::Spurious], 0);
 
         // 2. Lose packets (1, 2) --> `RecoveryStart`, 1 event, reduced cwnd
+        let cwnd_before_loss = cc.cwnd();
+        assert!(cc_stats.w_max.is_none());
```
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs]:
```suggestion
        assert_eq!(cc_stats.w_max, None);
```
You used this elsewhere... (also in other places)

File: neqo-transport/src/cc/classic_cc.rs [hunk 5]
```diff
@@ -1642,6 +1646,9 @@ mod tests {
             Some(SlowStartExitReason::CongestionEvent)
         );
         assert_eq!(cc_stats.congestion_events[CongestionEvent::Loss], 1);
+        #[expect(clippy::cast_precision_loss, reason = "cwnd_before_loss is small")]
+        let expected_w_max = cwnd_before_loss as f64;
+        assert_eq!(cc_stats.w_max, Some(expected_w_max));
```
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs]: `f64` comparison

**Review by larseggert — APPROVED** [review]

File: neqo-transport/src/cc/tests/cubic.rs (hunk 3 above)
Comment (omansfeld) [reply neqo-transport/src/cc/tests/cubic.rs:308]: Very interesting read, that part was amazing:

> If you haven’t memorized pi to 15+ digits then this may be lost on you, but the salient point is that float(pi) + sin(float(pi)) is a more accurate value of pi than float(pi). Or, alternately, sin(float(pi)) tells you nothing more than the error in float(pi).

I think here it just highlights the fundamental problem that we use `f64` for a bunch of values that represent bytes (so really should be `usize`) because they are used in floating point math in Cubic that then converts it's results back to `usize` when returning values to the outside. That is already tracked in #3022.

So really most of the comparisons in the cubic tests are comparing whole numbers (as `f64`) that have been assigned directly from `usize`. I guess that makes them less flaky. At least I have never noticed tests failing intermittently because of the comparisons in this file.

File: neqo-transport/src/cc/classic_cc.rs (hunk 4 above)
Comment (omansfeld) [reply neqo-transport/src/cc/classic_cc.rs]: Thanks for pointing it out, I fixed that in #3432 and missed it here.

File: neqo-transport/src/cc/classic_cc.rs (hunk 5 above)
Comment (omansfeld) [reply neqo-transport/src/cc/classic_cc.rs]: Converted to `usize` instead and added more explanations in the clippy reason. Also see the comment above about the whole `f64` situation.
